policy_wait: Add timeouts for total time and when an endpoint failed

jrajahalme opened the pull request:
cilium policy wait waits indefinitely if endpoint policy regeneration
fails. Add a timeout both for total time waited and for when an
endpoint regeneration fails. These default to 120 and 10 seconds,
respectively.

Fixes: #2070
Reported-by: Ian Vernon <ian@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

tgraf reviewed (changes requested):
I'm fine with this but the default should be higher. 360 seconds and 60 seconds.

joestringer commented on cilium/cmd/policy_wait.go:
```diff
@@ -48,30 +48,44 @@ var policyWaitCmd = &cobra.Command{
 	policyCmd.AddCommand(policyWaitCmd)
 	policyWaitCmd.Flags().IntVar(&waitTime, "sleep-time", 1, "Sleep interval between checks (seconds)")
+	policyWaitCmd.Flags().IntVar(&failWaitTime, "fail-wait-time", 60, "Wait time after which command fails if endpoint regeration fails (seconds)")
+	policyWaitCmd.Flags().IntVar(&maxWaitTime, "wait-time", 360, "Wait time after which command fails (seconds)")
```
Perhaps this should be `max-wait-time` to be more clear about its interaction with `fail-wait-time`

jrajahalme replied:
OK.

joestringer commented on cilium/cmd/policy_wait.go:
```diff
@@ -38,7 +38,7 @@ var policyWaitCmd = &cobra.Command{
 		}
 
-		haveWaited := false
+		timeWaited := 0
```
Did you consider expressing this more naturally using native time objects?

```
startTime := time.Now()
failDeadline := startTime.Add(failWaitTime * time.Second)
maxDeadline := startTime.Add(maxWaitTime * time.Second)
...
} else if time.Now().After(failDeadline) && notReady > 0 {
...
} else if time.Now().After(maxDeadline) {
```

This is a little more readable to me, and the log messages can also do something like "failed after %s ..", time.Since(startTime) and it will be more accurate.

jrajahalme replied:
Thanks for the ready-to-use suggestion :-)
Done.

joestringer commented on cilium/cmd/policy_wait.go:
```diff
@@ -48,30 +48,44 @@ var policyWaitCmd = &cobra.Command{
 				}
 				return
+			} else if timeWaited > failWaitTime && notReady > 0 {
+				fmt.Printf("\n")
```
Any particular reason why these newlines don't fold into the `Fatalf()` call?

jrajahalme replied:
No. Had to check they go to the same stream and they do.

jrajahalme replied:
I.e., folded, thanks!

joestringer reviewed (commented):
I have some minor usability feedback on this which should be easy to address.

jrajahalme committed:
policy_wait: Add timeouts for total time and when an endpoint failed

cilium policy wait waits indefinitely if endpoint policy regeneration
fails. Add a timeout both for total time waited and for when an
endpoint regeneration fails. These default to 120 and 10 seconds,
respectively.

Fixes: #2070
Reported-by: Ian Vernon <ian@covalent.io>
Signed-off-by: Jarno Rajahalme <jarno@covalent.io>
